### PR TITLE
Victor/adding more abstractions

### PR DESCRIPTION
**Pull request by victorsowa12**
Bring over some utc -> datetime and datetime -> utc conversion functions. Also add some more time remaining functions. The utc/datetime functions are used by bcmp_time.

### CODE REVIEW

File: include/bm_os.h [hunk 1]
```diff
@@ -34,21 +34,38 @@ BmErr bm_semaphore_give(BmSemaphore semaphore);
...
+#define time_remaining_ticks_from_ISR(startTicks, timeoutTicks)                \
+  time_remaining(startTicks, bm_get_tick_count_from_isr(), timeoutTicks)
+#define time_remaining_ms(startTimeMs, timeoutMs)                              \
+  bm_ticks_to_ms(time_remaining(bm_ms_to_ticks(startTimeMs),                   \
+                                bm_get_tick_count(),                           \
+                                bm_ms_to_ticks(timeoutMs)))
+#define time_remaining_ms_from_ISR(startTimeMs, timeoutMs)                     \
+  bm_ticks_to_ms(time_remaining(bm_ms_to_ticks(startTimeMs),                   \
+                                bm_get_tick_count_from_isr(),                  \
+                                bm_ms_to_ticks(timeoutMs)))
```
Comment (victorsowa12) [inline include/bm_os.h:67]: added these and `bm_get_tick_count_from_isr`, the rest of the changes in this file are formatting.

File: src/util.c [hunk 2]
```diff
@@ -35,3 +35,115 @@ void swap_64bit(void *x) {
...
+/*!
+  Convert unix timestamp(UTC) in microseconds to date (YYYYMMDDhhmmss)
+
+  \param[in] utc_us - utc in microseconds
+  \param[out] UtcDateTime - date time
+  \return None
+*/
+void date_time_from_utc(uint64_t utc_us, UtcDateTime *date_time) {
+  // TODO - Do we really need this?
+  // configASSERT(dateTime);
```
Comment (victorsowa12) [inline src/util.c]: Yeah, still not sure we really need this assert + haven't come up with a standard way for someone to include/exclude asserts yet.

File: include/util.h [hunk 3]
```diff
@@ -39,9 +39,24 @@ typedef enum {
...
+  uint32_t usec;
+} UtcDateTime;
+
 #define array_size(x) (sizeof(x) / sizeof(x[0]))
 #define member_size(type, member) (sizeof(((type *)0)->member))
 
+uint32_t time_remaining(uint32_t start, uint32_t current, uint32_t timeout);
+uint32_t utc_from_date_time(uint16_t year, uint8_t month, uint8_t day,
+                            uint8_t hour, uint8_t minute, uint8_t second);
+void date_time_from_utc(uint64_t utc_us, UtcDateTime *date_time);
```
Comment (victorsowa12) [inline include/util.h:58]: Copied these over from app_util.h/util.c in bm_protocol + all of the help functions for leap years etc. I renamed them to match the code style guidelines.

File: src/util.c (hunk 2 above)
Comment (matt001k) [reply src/util.c]: It could be a simple check to avoid the assert for now:
```
if (data_time) {
...
}
```

File: src/util.c [hunk 4]
```diff
@@ -35,3 +35,115 @@ void swap_64bit(void *x) {
...
+
+// leap year calulator expects year argument as years offset from 1970
+#define LEAP_YEAR(Y)                                                           \
+  (((1970 + Y) > 0) && !((1970 + Y) % 4) &&                                    \
+   (((1970 + Y) % 100) || !((1970 + Y) % 400)))
+
+#define SECS_PER_MIN (60UL)
+#define SECS_PER_HOUR (3600UL)
+#define SECS_PER_DAY (SECS_PER_HOUR * 24UL)
+#define MICROSECONDS_PER_SECOND (1000000)
```
Comment (matt001k) [inline src/util.c]: These macros should be `lower_snake_case`.

**Review by matt001k — APPROVED** [review]
Nice work, couple of comments but everything looks good.

File: src/util.c (hunk 2 above)
Comment (victorsowa12) [reply src/util.c]: I like that idea! Went with it!

**Pull request merged**